mlflow/pyfunc/__init__.py

```diff
@@ -1054,6 +1127,46 @@ def spark_udf(spark, model_uri, result_type="double", env_manager=_EnvManager.LO
 
     model_metadata = Model.load(os.path.join(local_model_path, MLMODEL_FILE_NAME))
 
+    model_output_schema = model_metadata.get_output_schema()
+
+    if result_type is None:
+        if model_output_schema is None:
+            _logger.warning(
+                "No 'result_type' provided for spark_udf and the model metadata does not "
+                "have an output schema. 'result_type' is set to 'double' type."
+            )
+            result_type = "double"
+        else:
+            result_type = _infer_spark_udf_return_type(model_output_schema)
+    else:
+        result_type = "boolean" if result_type == "bool" else result_type
+
+    if not isinstance(result_type, SparkDataType):
+        result_type = _parse_datatype_string(result_type)
+
+    elem_type = result_type
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+        if isinstance(elem_type, ArrayType):
+            elem_type = elem_type.elementType
+
+    supported_types = [
+        IntegerType,
+        LongType,
+        FloatType,
+        DoubleType,
+        StringType,
+        BooleanType,
+        SparkStructType,
+    ]
+
+    if not any(isinstance(elem_type, x) for x in supported_types):
+        raise MlflowException(
+            message="Invalid result_type '{}'. Result type can only be one of or an array of one "
+            "of the following types: {}".format(elem_type, supported_types),
+            error_code=INVALID_PARAMETER_VALUE,
+        )
```

Comment:
if `elem_type` is ArrayType because `result_type` is a triple-nested array, this error message is confusing. It'll say "Result type can can only be an array of one of the following types: IntegerType, LongType, FloatType , etc.`

But this isn't true, it's okay for the result type to be an array of array of int, etc. just not array of array of array of int. 

I think we need a separate exception

Reply:
I will refactor this part code. Current code does not check all cases

Reply:
Related code is refactored :)  See `_check_udf_return_type` function.